internal/task: use asyncify on WASM

niaow opened the pull request:
This currently requires manually running the compiled wasm through `wasm-opt`.
Very little works so far.
Issue: #1101

niaow committed:
internal/task: use asyncify on webassembly

This change implements a new "scheduler" for WebAssembly using binaryen's asyncify transform.
This is more reliable than the current "coroutines" transform, and works with non-Go code in the call stack.

runtime (js/wasm): handle scheduler nesting

If WASM calls into JS which calls back into WASM, it is possible for the scheduler to nest.
The event from the callback must be handled immediately, so the task cannot simply be deferred to the outer scheduler.
This creates a minimal scheduler loop which is used to handle such nesting.

dgryski commented:
What work is remaining to get this PR finished?  I'd love to help out any way I can.

niaow commented:
This is mainly blocked by #2035 I think

dgryski commented:
While that bug is getting fixed, is there work that could be done on this issue even with having to run wasm-opt manually?

aykevl commented:
I'm almost inclined to say that we should perhaps just get this finished (even without #2035). That means compiling for WebAssembly requires an extra binary - but I think that's only required when using the new Asincify based scheduler?

Thoughts?
I don't like adding an external dependency (it makes distribution a whole lot harder) but right now concurrency on WebAssembly is already somewhat broken and this PR might fix it.

niaow commented:
Alright, I can modify this to just shell out to wasm-opt for now when using asyncify.

niaow commented:
I made that change and attempted to debug some issues. It appears that after the initial launch suspend, resumed goroutines will not suspend again. I don't quite understand it yet.

dgryski commented:
I'll take a look at this on Monday and see what I can come up with.

niaow commented:
I have fixed what was probably the biggest bug. However, there are still some things left to be fixed apparently, such as weird WASI crashes and some sort of unicode error in the reflect test (. . . and i still havent added binaryen to CI).

dgryski commented:
\o/. This works great on my little test `ping pong` program (included below)

```
package main

import (
	"time"
)

func pinger(pinger <-chan int, ponger chan<- int, who string) {
	for {
		<-pinger
		println(who)
		time.Sleep(time.Second)
		ponger <- 1
	}
}

func main() {
	ping := make(chan int)
	pong := make(chan int)

	go pinger(ping, pong, "ping")
	go pinger(pong, ping, "pong")

	ping <- 1

	select {}
}
```

However, I'm getting an unaligned pointer error with `gc=leaking` on `wasmtime` when trying to read the clock.

I managed to fix it with this patch.

```
 func ticks() timeUnit {
-       var nano uint64
-       clock_time_get(0, timePrecisionNanoseconds, &nano)
+       var b [16]byte
+       ptr := uintptr(unsafe.Pointer(&b[0]))
+       ptr += 8 + (^ptr & 0x7) + 1
+
+       clock_time_get(0, timePrecisionNanoseconds, (*uint64)(unsafe.Pointer(ptr)))
+       nano := *(*uint64)(unsafe.Pointer(ptr))
        return timeUnit(nano)
 }

```

It seems like a proper fix lies elsewhere, although I'm not sure where..

Trying to run the above ping-pong code with `gc=conservative` instead produces:

```
Caused by:
    0: failed to invoke command default
    1: wasm trap: uninitialized element
       wasm backtrace:
           0: 0x6c376 - <unknown>!<wasm function 266>
           1: 0x1489c - <unknown>!<wasm function 68>
           2: 0x2e336 - <unknown>!<wasm function 158>
           3: 0x2d77d - <unknown>!<wasm function 156>
           4: 0x2d636 - <unknown>!<wasm function 155>
```

I'm still looking into this..

niaow commented:
> However, I'm getting an unaligned pointer error with `gc=leaking` on `wasmtime` when trying to read the clock.

I suspect this may be fixed by #2009. This PR is kinda old so I can rebase tommorow.

dgryski commented:
Adding `-g` to the options provided to `wasm-opt` with the following patch:

```
diff --git a/builder/build.go b/builder/build.go
index df329bd7..7568e8ae 100644
--- a/builder/build.go
+++ b/builder/build.go
@@ -641,7 +641,7 @@ func Build(pkgName, outpath string, config *compileopts.Config, action func(Buil
                                default:
                                        return fmt.Errorf("unknown opt level: %q", config.Options.Opt)
                                }
-                               cmd := exec.Command("wasm-opt", "--asyncify",
+                               cmd := exec.Command("wasm-opt", "-g", "--asyncify",
                                        "--optimize-level", strconv.Itoa(optLevel),
                                        "--shrink-level", strconv.Itoa(shrinkLevel),
                                        executable, "--output", executable)
```

gives me a much better error:
```
Caused by:
    0: failed to invoke command default
    1: In func wasi_snapshot_preview1::args_sizes_get at write size: Pointer not aligned to 4: Region { start: 1066183, len: 4 }
       wasm backtrace:
           0: 0x320f - <unknown>!runtime.initAll
           1: 0x307f - runtime.run$1
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler_any.go:19:10
           2: 0x301d - <goroutine wrapper>
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler_any.go:18:2
           3: 0x3cb6 - <unknown>!tinygo_launch
           4:  0x538 - (*internal/task.Task).Resume
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/internal/task/task_asyncify.go:94:17
           5: 0x2edb - runtime.scheduler
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler.go:171:11
                     - runtime.run
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler_any.go:24:11
                     - _start
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/runtime_wasm_wasi.go:21:5

```

So indeed, something about asyncify is futzing with the alignment of the pointers and making things unhappy.

dgryski commented:
I rebased my local copy against `dev` with no success.  Same pointer alignment issue.  I'll keep looking at this..

dgryski commented:
Patching gc_leaking.go with
```
 func initHeap() {
-       // preinit() may have moved heapStart; reset heapptr
-       heapptr = heapStart
+       // preinit() may have moved heapStart; reset heapptr and make sure it's aligned
+       heapptr = align(heapStart)
 }
```

allows both gc=conservative and gc=leaking to run the pingpong example.

niaow commented:
For some reason I thought chocolatey had binaryen. I guess I will need to switch over the install then.

dgryski commented:
Something about the asyncify code is causing https://github.com/tinygo-org/tinygo/issues/2172 to fail

aykevl commented on azure-pipelines.yml:
```diff
@@ -15,6 +15,19 @@ jobs:
+          # install ninja
+          choco install ninja
+          # hack ninja to use fewer jobs
+          echo -e 'C:\\ProgramData\\Chocolatey\\bin\\ninja -j4 %*' > /usr/bin/ninja.bat
```
There is no reason to do this when not building LLVM.

aykevl commented on compiler/compiler_test.go:
```diff
@@ -93,6 +101,7 @@ func TestCompiler(t *testing.T) {
 				Scheduler:          config.Scheduler(),
 				FuncImplementation: config.FuncImplementation(),
 				AutomaticStackSize: config.AutomaticStackSize(),
+				DefaultStackSize:   4096,
```
I think this should match the target JSON for consistency.

```suggestion
				DefaultStackSize:   config.DefaultStackSize(),
```

aykevl commented on src/internal/task/task_asyncify.go:
Normally, I would expect an aligned stack to be passed in. For example, a stack size of 2048 is very much aligned.
I never really thought about aligning it, but yeah, technically it might need to be aligned.

I don't think this needs to be done at runtime (although there is a good chance it's optimized away in this case): it's only ever specified in code (in compileopts/target.go for example) or in target JSON files. The only way to easily change it is by modifying the JSON file directly, for example. If we're going to align it, I think it would be better to do it at compile time instead of at runtime, for example here:

https://github.com/tinygo-org/tinygo/blob/157382600535c95bdbf5d6e8c4b9e5ca71749018/compiler/goroutine.go#L125

This can be done by aligning to 16 bytes, I believe that's nowadays a pretty standard alignment (even on riscv32 and i386, for example).

aykevl commented on src/internal/task/task_asyncify.go:
So I guess once this call returns, we're back after rewinding?

niaow replied:
Yes.

aykevl commented on targets/wasi.json:
```diff
@@ -5,6 +5,9 @@
 	"linker":        "wasm-ld",
 	"libc":          "wasi-libc",
+	"scheduler":     "asyncify",
+	"gc":            "conservative",
```
This GC was already the default, so this is not technically needed. But many other targets also do it so :shrug:

aykevl commented on testdata/channel.go:
```diff
@@ -244,7 +244,7 @@ func receive(ch <-chan int) {
 	for i := 1; i <= 8; i++ {
 		if i == 4 {
-			time.Sleep(time.Microsecond)
+			time.Sleep(time.Second)
```
Why did you increase this to a second? That's some valuable time spent in a test. If you want to make it longer than a microsecond, please consider making it multiple microseconds or even a whole millisecond.

niaow replied:
This was a temporary change I made when debugging a timer queue crash. I will revert this when cleaning up.

aykevl reviewed (commented):
This looks pretty clean. I would like to take a more closer look at some parts, but so far I only have some relatively minor comments.

aykevl commented:
> For some reason I thought chocolatey had binaryen. I guess I will need to switch over the install then.

`scoop` has it, not `choco`.

niaow commented:
I have created a fix for the WASM scheduler recursion failure: #2178

dgryski commented:
I think this branch still has an odd failure.  Trying to build the following program with
`tinygo build -scheduler=asyncify -gc=leaking -target=wasi -o m.wasm main.go`


```
package main

import (
	"context"
)

func main() {
	ctx := context.Background()

	select {
	case <-ctx.Done():
	}
}
```

produces the following build failure:
```
Call parameter type does not match function signature!
%runtime.channel.56* null
 %runtime.channel.80*  %1 = call i1 bitcast (i1 (%runtime.channel*, i8*, %runtime.channelBlockedList*, i8*, i8*)* @runtime.chanRecv to i1 (%runtime.channel.80*, i8*, %runtime.channelBlockedList.79*, i8*, i8*)*)(%runtime.channel.56* null, i8* null, %runtime.channelBlockedList.79* %chan.blockedList, i8* undef, i8* null), !dbg !1627
```

niaow commented:
I have been poking around, and it currently appears that the interface lowering pass is the first pass introducing type errors when building that code.

niaow commented:
I identified the issue and seperated it into #2197

niaow commented:
Hmm, it looks like LLVM 10 may have issues with the not-yet-resolved symbols for asyncify. I am not really sure how we want to handle this. I guess we could disable this on LLVM 10? I am not sure of a better option.

dgryski commented:
Given that llvm 13 is out now, how many versions behind is tinygo committed to supporting?

aykevl commented:
> Hmm, it looks like LLVM 10 may have issues with the not-yet-resolved symbols for asyncify. I am not really sure how we want to handle this. I guess we could disable this on LLVM 10? I am not sure of a better option.

Don't let that be a blocker. If LLVM 10 is an issue, we'll drop LLVM 10 support. That's a really old version by now. (Usually we support the last two LLVM versions or so, to make things easier for distributions). Even more so because this PR is a blocker to add support for newer LLVM versions.

aykevl commented:
I've created #2216 to drop support for LLVM 10 so that this PR can continue without being burdened by that LLVM version.

niaow commented:
Alright, sounds good. I will take a look at it tomorrow.
The main other issues I still need to deal with right now are:
1. Fix Windows CI
2. Figure out why `test-llvm11-go116` is timing out

aykevl commented on azure-pipelines.yml:
```diff
@@ -15,6 +15,16 @@ jobs:
+          # install qemu
+          choco install qemu --version=2020.06.12
+          # install binaryen
+          wget https://github.com/WebAssembly/binaryen/releases/download/version_101/binaryen-version_101-x86_64-windows.tar.gz
```
Apparently wget isn't included on this Windows version: https://github.com/actions/virtual-environments/blob/main/images/win/Windows2016-Readme.md
You may have better luck with curl, or by installing emscripten using Scoop instead: ScoopInstaller/Main#2740 (installing Scoop should be pretty fast).

deadprogram commented:
Now that #2216 has been merged into `dev` branch, this PR should be able to continue. :smile_cat:

niaow commented:
So it appears that we may need to require binaryen version 102 instead of 101. This makes it a bit more annoying to install, since 102 does not have pre-built binaries.

aykevl commented:
> So it appears that we may need to require binaryen version 102 instead of 101.

Are you referring to this failure?

```
build/tinygo build             -o test.wasm -tags=hifive1b             examples/blinky1
error: wasm-opt failed: signal: segmentation fault (core dumped)
```

niaow commented:
> Are you referring to this failure?

Yeah, I think that is fixed in binaryen 102. I am not 100% sure, although it seems to match up with a bug report in 101, and I have not been able to replicate any similar fault on 102.

deadprogram commented:
Any action on this PR? Quite a few people are eagerly awaiting it (see what I did there?).

niaow commented:
> Any action on this PR? Quite a few people are eagerly awaiting it (see what I did there?).

Sorry I haven't gotten a chance to work on it this week.
Tomorrow I should hopefully be able to get back to it.
I am going to try to get all of the targets to build binaryen 102 from source and bundle it in the release tarball.

niaow commented:
Getting this to work on stretch is seemingly more difficult than I anticipated, due to the requirement for a newer cmake version.

aykevl commented:
Did you verify that the bug is indeed there in version 101 and fixed in 102? It could be something unrelated to the version.

aykevl commented:
With version 101:

```
$ $WASMOPT --version
wasm-opt version 101 (version_101)

$ tinygo build -o test.wasm -tags=hifive1b -size=short examples/blinky1
   code    data     bss |   flash     ram
  29870     771  130301 |   30641  131072
```

I'm compiling version 102 right now.

niaow commented:
> Did you verify that the bug is indeed there in version 101 and fixed in 102? It could be something unrelated to the version.

It is a non-deterministic bug, and does not replicate on my system. It doesn't seem to replicate on 102 so far, so it may be right it may be wrong but cant tell for sure.

aykevl commented:
Hmm, perhaps with was a flaky test then?
If it can't be reproduced, I'd say don't worry about bundling wasm-opt in all TinyGo releases.

niaow commented:
Not a flaky test. This is a legitimate bug, and reproduces randomly in CI (generally at least one of the builds fails, but rarely more than that) with 101.

niaow commented:
I think(?) I have fixed all of the build errors at this point. I just need to wait for all of the CI run (and Windows may take a while).

niaow commented:
I had removed the `rm -rf llvm-source` without realizing why it was done. I updated the comments in the CI config to try to explain it better, because it was surprising behavior.

deadprogram commented:
All tests are passing now!

niaow commented:
@deadprogram windows smoke tests are still running: https://github.com/niaow/tinygo/runs/4132134414?check_suite_focus=true

niaow commented:
Windows smoke tests failed.

niaow commented:
It looks like I messed up the autodetection on Windows (I need to add a `.exe` suffix).

niaow commented:
The commit I just pushed should fix it, although it will take 2 hours to find out. I think this is ready for review to start now.

aykevl commented on .circleci/config.yml:
```diff
@@ -25,7 +25,9 @@ commands:
                 gcc-avr \
-                avr-libc
+                avr-libc \
+                cmake \
+                ninja-build
```
Can you do this installation only when these tools are needed in the build-binaryen step? That speeds up the CI in the common case.

niaow replied:
I did this like this for 2 reasons:
1. We need to install these differently on buster and stretch.
2. We need these for both LLVM and Binaryen, and avoiding a double install seemed annoying.

I guess I could potentially restore both caches before installing, although it seems to make this build config a lot more complicated.

niaow replied:
This appears to only make CI 1 second slower. I don't think the extra complexity is worthwhile.

aykevl replied:
> This appears to only make CI 1 second slower. I don't think the extra complexity is worthwhile.

How did you measure this? It seems rather unlikely that installing software is that fast.
(If it is really just 1 second, then installing it unconditionally would be fine).

niaow replied:
I looked at 2 a build on the `dev` branch and a build from this PR. The "install apt dependencies steps" took 22 seconds on this PR and 23 seconds on the dev branch. Looking through more CI runs on `dev` now I see a lot more variation, but the difference in download size is 180 MB to 186 MB which gives me a seemingly appropriate estimate of 3%. These packages are relatively small compared to the other packages which we install.

aykevl replied:
Ok. Well maybe it is fine then. We can revisit this later, especially when we move away from Debian stretch.

aykevl commented on .circleci/config.yml:
```diff
@@ -141,9 +175,14 @@ commands:
                 gcc-avr \
-                avr-libc
+                avr-libc \
+                ninja-build \
+                python3
```
Same here.

aykevl commented on .circleci/config.yml:
```diff
@@ -206,9 +240,14 @@ commands:
                 gcc-avr \
-                avr-libc
+                avr-libc \
+                ninja-build \
+                python3
```
And here.

aykevl commented on .circleci/config.yml:
```diff
@@ -283,7 +317,7 @@ commands:
             sudo tar -C /usr/local -xzf go1.17.darwin-amd64.tar.gz
             ln -s /usr/local/go/bin/go /usr/local/bin/go
-            HOMEBREW_NO_AUTO_UPDATE=1 brew install qemu
+            HOMEBREW_NO_AUTO_UPDATE=1 brew install qemu cmake ninja
```
And here.

aykevl commented on .circleci/config.yml:
```diff
@@ -76,6 +85,29 @@ commands:
+          command: |
+            echo -e '#!/bin/sh\n/usr/bin/ninja -j3 "$@"' > /go/bin/ninja
+            chmod +x /go/bin/ninja
+  build-binaryen:
```
Suggestion: this might be clearer when the name is `build-binaryen-linux`. Because the cache is only for Linux.
Similarly, I think it's better to rename the cache key to something like `binaryen-linux-v1`.

aykevl commented on .gitignore:
```diff
@@ -1,4 +1,3 @@
-build
```
Oh, I see you've moved the .gitignore line into a separate file in the build directory.
I guess that's fine, just wondering what the reason is?

niaow replied:
I needed to keep the `build` directory, and for some reason was struggling to get it to track as-is (negated patterns seemed to be behaving weirdly). This is just how I normally do that.

aykevl commented on Makefile:
```diff
@@ -169,6 +169,12 @@ $(LLVM_BUILDDIR)/build.ninja: llvm-source
+.PHONY: binaryen
+binaryen: build/wasm-opt
+build/wasm-opt:
+	cd lib/binaryen; cmake -G Ninja . -DBUILD_STATIC_LIB=ON; ninja
```
```suggestion
	cd lib/binaryen && cmake -G Ninja . -DBUILD_STATIC_LIB=ON && ninja
```

Not sure whether this works as-is in a Makefile but generally not continuing on errors is a good thing.

aykevl reviewed (commented):
I've left a few comments about how the CI can be sped up a bit, but in general this looks very good.

I have one question though: what happens if you export a function and do a blocking call in there? JavaScript can't block, so it must return _something_ instead. This could be surprising to users, especially if this behavior is not part of the function signature.

aykevl commented:
> The commit I just pushed should fix it, although it will take 2 hours to find out. I think this is ready for review to start now.

Yeah, caching on GitHub Actions is a bit weird. It should work, but for some reason there are a lot of cache misses.

niaow commented:
> I have one question though: what happens if you export a function and do a blocking call in there? JavaScript can't block, so it must return something instead. This could be surprising to users, especially if this behavior is not part of the function signature.

Assuming that you use `js.FuncOf`. . .  the current behavior (unchanged by this PR) is that stuff will break. Go defines this to deadlock, but our implementation is still non-conformant (I hope to fix this in a future PR).

As for raw exported functions, they will panic in `task.Pause` if there are no running goroutines. If there are running goroutines (Go calls into JS calls into Go which then blocks), this will break and bad things will happen. This particular issue is a lot more difficult to solve, and will require us to provide our own implementation of `syscall/js` with some changes.

aykevl commented:
> As for raw exported functions, they will panic in `task.Pause` if there are no running goroutines. If there are running goroutines (Go calls into JS calls into Go which then blocks), this will break and bad things will happen.

Right. That was what I was wondering about. I suspect many people today call exported functions from TinyGo, as described [here](https://tinygo.org/docs/guides/webassembly/). Maybe `-scheduler=none` is the solution in that case, but maybe that's a bit too drastic. In any case, we have to provide some answer to the question of how to call into Go functions from JavaScript.
(This is why I prefer to keep `-scheduler=coroutines` around for one release if possible).

niaow commented:
Something seems really wrong with the Windows build, so that might take a bit to debug. https://github.com/niaow/tinygo/runs/4132565997?check_suite_focus=true#step:17:72

(This means that binaryen crashed when just run with `--version`)

aykevl commented:
@niaow you might want to rebase on top of the dev branch, #2245 should avoid rebuilding LLVM on each push.

niaow commented:
It looks like the version of GCC used by Debian Stretch is too old to build `binaryen` correctly, and that the cache was hiding this fact.

niaow commented:
I managed to work around that by using the copy of clang that we build.
Now the only remaining issue is Windows.

niaow commented:
If my debugging attempts are correct, it looks like binaryen is still trying to dynamically link against `libstdc++` and `libgcc` from mingw, and failing (error code `0xc0000139`).
```
[niaow@finch bin]$ objdump -p wasm-opt.exe | grep "DLL Name:"
	DLL Name: libgcc_s_seh-1.dll
	DLL Name: KERNEL32.dll
	DLL Name: msvcrt.dll
	DLL Name: libwinpthread-1.dll
	DLL Name: USER32.dll
	DLL Name: libstdc++-6.dll
```

aykevl commented:
For the TinyGo build I use `-static -static-libgcc -static-libstdc++`, I think you might need those for binaryen too.

niaow commented:
Yeah I just realized that. Adding now.

aykevl commented:
I can try rebooting to Windows to test some build options, would that help?

niaow commented:
I suspect that I just fixed it. If not, I finally have a Windows VM setup so I can test there.

niaow commented:
All tests are now passing.

niaow commented:
I did a quick squash/cleanup, so I think this is ready for a final review.

deadprogram commented:
Do we need to modify the dev Dockerfile if you want to use this scheduler? https://github.com/tinygo-org/tinygo/blob/dev/Dockerfile#L23-L28

deadprogram commented on Dockerfile:
```diff
@@ -30,7 +30,7 @@ COPY --from=tinygo-base /tinygo/targets /tinygo/targets
     apt-get update && \
     apt-get install -y make clang-11 libllvm11 lld-11 && \
-    make wasi-libc
+    make wasi-libc binaryen
```
Oops, just noticed this. Never mind my other question.

niaow replied:
I just double checked and this doesn't actually work anymore, so I do actually need to fix it.

deadprogram commented on Makefile:
```diff
@@ -54,6 +54,8 @@ ifeq ($(OS),Windows_NT)
     CGO_LDFLAGS += -static -static-libgcc -static-libstdc++
     CGO_LDFLAGS_EXTRA += -lversion
 
+	BINARYEN_OPTION += -DCMAKE_EXE_LINKER_FLAGS='-static-libgcc -static-libstdc++'
```
I'm almost embarrassed to point out something so trivial as an extra tab here :joy_cat:

niaow replied:
Huh. I guess this is what happens when using 4-wide tabs in editor :facepalm:

niaow committed:
fix binaryen build in docker

Oops I forgot to install cmake and ninja.

deadprogram reviewed (approved):
All tests are now passing, and seems good to me.

deadprogram commented:
@aykevl any last comments before merge?

gedw99 commented:
Is this the fix to get WASM compilation working again ?

deadprogram commented:
> Is this the fix to get WASM compilation working again ?

WASM compilation is working afaik. This PR is to add asyncify as an option for task scheduling for WASM programs.

gedw99 commented:
> > Is this the fix to get WASM compilation working again ?
> 
> WASM compilation is working afaik. This PR is to add asyncify as an option for task scheduling for WASM programs.

ok thank for the response. 

https://github.com/tinygo-org/tinygo/issues/1972 is blocking i believe. I think is for gio when building to Web ( wasm ).

aykevl reviewed (commented):
This looks good to me.
I can't really comment on the Dockerfile changes, is the issue there fixed now? I don't know a whole lot about Docker.

aykevl commented on goenv/goenv.go:
```diff
@@ -67,11 +70,98 @@ func Get(name string) string {
 }
 
+// Find wasm-opt, or exit with an error.
+func findWasmOpt() string {
```
[this comment was intended to be part of the review but I forgot to click the "start review" button]

I think this function can be simplified. In short: the bundled wasm-opt binaries don't need to be checked for a version.

I would suggest the following logic:

```
if $TINYGOROOT/bin/wasm-opt[.exe] exists:
    return this path
if $TINYGOROOT/build/wasm-opt[.exe] exists:
    return this path
look for wasm-opt using LookPath
if not found:
    print or return error
check for version
if version too old:
    return error
return this path
```

This is not a blocker to me.

niaow replied:
I think we should still be doing a version check regardless for now, until we can actually link it in properly.

niaow replied:
This has already caught one issue, so it is still helpful.

niaow commented:
The Dockerfile issues have been fixed.

deadprogram commented:
OK merge time! Great work on this @niaow thank you. Also thanks to everyone who helped review/test.
